### PR TITLE
Make the service worker send custom response

### PR DESCRIPTION
**Pull request by creativcoder**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] These changes do not require tests because "refactor".

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12582)

### CODE REVIEW

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r1.
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/abstractworker.rs, line 21 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNXHVxfdqPuS3OhHd3N:-KNXHVxfdqPuS3OhHd3O:2044650566) ([raw file](https://github.com/servo/servo/blob/00db4bbb6cd8c10416810ad528f7ead819744a97/components/script/dom/abstractworker.rs#L21)):*
> ```Rust
>     DOMMessage(StructuredCloneData),
>     // custom response by service worker
>     Response(CustomResponseMediator)
> ```

Could create a separate ServiceWorkerScriptMsg that includes this and a WorkerScriptMsg variant? I'm not sure about all the places where we store senders and receivers, but it would be nice to avoid introducing another message that's only used by one worker type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582#-:-KNXI9jghEAMBjUsyNcw:1458266033)*

**Comment by jdm** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582)*

**Comment by creativcoder** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/abstractworker.rs, line 21 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNXHVxfdqPuS3OhHd3N:-KNbMQ-SJeXE9E4IClrl:727549967) ([raw file](https://github.com/servo/servo/blob/00db4bbb6cd8c10416810ad528f7ead819744a97/components/script/dom/abstractworker.rs#L21)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Could create a separate ServiceWorkerScriptMsg that includes this and a WorkerScriptMsg variant? I'm not sure about all the places where we store senders and receivers, but it would be nice to avoid introducing another message that's only used by one worker type.
</details>
okay, so I've added the `ServiceWorkerScriptMsg` enum as a seperate msg type in `serviceworkerglobalscope.rs` itself. Does that look ok to you ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582)*

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/serviceworker_manager.rs, line 153 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNbQkzGN6ZBe61LFfiH:-KNbQkzGN6ZBe61LFfiI:1028971453) ([raw file](https://github.com/servo/servo/blob/95cb25832ab6bc83c3a31e9912d492ef3f48c2ec/components/script/serviceworker_manager.rs#L153)):*
> ```Rust
>     fn handle_message_from_resource(&mut self, mediator: CustomResponseMediator) -> bool {
>         let worker_sender = self.prepare_activation(&mediator.load_url);
>         if serviceworker_enabled() {
> ```

This should probably guard the `prepare_activation` call as well.

---

*[components/script/dom/serviceworkerglobalscope.rs, line 44 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNbQbJl-AqUCDv4T4lt:-KNbQbJl-AqUCDv4T4lu:1514451540) ([raw file](https://github.com/servo/servo/blob/95cb25832ab6bc83c3a31e9912d492ef3f48c2ec/components/script/dom/serviceworkerglobalscope.rs#L44)):*
> ```Rust
>     Common(CommonScriptMsg),
>     /// Message sent through postMessage
>     DOMMessage(StructuredCloneData),
> ```

What if we replace these two variants with a `CommonWorker(WorkerScriptMsg)` variant?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582#-:-KNbQqYu_T0twdCjYBph:1458266033)*

File: components/script/dom/serviceworkerglobalscope.rs [hunk 1]
```diff
@@ -223,19 +231,22 @@ impl ServiceWorkerGlobalScope {
...
+            ServiceWorkerScriptMsg::CommonWorker(WorkerScriptMsg::Common(CommonScriptMsg::RunnableMsg(_, runnable))) => {
                 runnable.handler()
             },
-            WorkerScriptMsg::Common(CommonScriptMsg::RefcountCleanup(addr)) => {
+            ServiceWorkerScriptMsg::CommonWorker(WorkerScriptMsg::Common(CommonScriptMsg::RefcountCleanup(addr))) => {
                 LiveDOMReferences::cleanup(addr);
             },
-            WorkerScriptMsg::Common(CommonScriptMsg::CollectReports(reports_chan)) => {
+            ServiceWorkerScriptMsg::CommonWorker(WorkerScriptMsg::Common(CommonScriptMsg::CollectReports(reports_chan))) => {
                 let scope = self.upcast::<WorkerGlobalScope>();
```
Comment (creativcoder) [inline components/script/dom/serviceworkerglobalscope.rs]: @jdm I've made that change adding `CommonWorker`, but i think this becomes very nested in matching in `handle_script_event`. Also not sure to reformat the match arms as it exceeds tidy's line length check. Suggestions ?

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/serviceworkerglobalscope.rs, line 241 \[r4\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNWIQchjo6J-TXCni5b-r4-241:-KNbiUTd12IFEJtFN5I1:550575327) ([raw file](https://github.com/servo/servo/blob/cfff732608a59fb885ac39a910adfad60d1fa723/components/script/dom/serviceworkerglobalscope.rs#L241)):*
<details><summary><i>Previously, creativcoder (Rahul Sharma) wrote…</i></summary>
> @jdm I've made that change adding `CommonWorker`, but i think this becomes very nested in matching in `handle_script_event`. Also not sure to reformat the match arms as it exceeds tidy's line length check. Suggestions ?
</details>
Go ahead and `use ServiceWorkerScriptMsg::*` and any others that are necessary to reduce the line length.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582#-:-KNbif3BYcfYhZXvjaol:1458266033)*

**Comment by creativcoder** [comment]
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/serviceworkerglobalscope.rs, line 241 \[r4\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNWIQchjo6J-TXCni5b-r4-241:-KNbxgin_zLYmBf5vg7H:676635175) ([raw file](https://github.com/servo/servo/blob/cfff732608a59fb885ac39a910adfad60d1fa723/components/script/dom/serviceworkerglobalscope.rs#L241)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Go ahead and `use ServiceWorkerScriptMsg::*` and any others that are necessary to reduce the line length.
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582)*

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/serviceworkerglobalscope.rs, line 44 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNcCBiavUoUa3Ve_Tdh:-KNcCBiavUoUa3Ve_Tdi:-1322090294) ([raw file](https://github.com/servo/servo/blob/bad984af207cb4a9a0e41ef035fb1e8a839c30e0/components/script/dom/serviceworkerglobalscope.rs#L44)):*
> ```Rust
>     Common(CommonScriptMsg),
>     /// Message sent through postMessage
>     DOMMessage(StructuredCloneData),
> ```

Why did you undo this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582#-:-KNcCJuSpyhQHjA2Nv_5:1458266033)*

**Comment by creativcoder** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/serviceworkerglobalscope.rs, line 44 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12582#-KNcCBiavUoUa3Ve_Tdh:-KNcLUmKPFK7lJzv-TtS:-1880756388) ([raw file](https://github.com/servo/servo/blob/bad984af207cb4a9a0e41ef035fb1e8a839c30e0/components/script/dom/serviceworkerglobalscope.rs#L44)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Why did you undo this change?
</details>
Whoops ! Updated that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582)*

**Comment by jdm** [comment]
@bors-servo: r+

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12582#-:-KNcTDwvfndSIoxL5Gb1:-317921516)*

**Comment by jdm** [comment]
@creativcoder Why'd you close the PR?

**Comment by creativcoder** [comment]
@jdm Oh, my bad the closed button got clicked. Can we re-open this ?

**Comment by KiChjang** [comment]
@bors-servo retry
* The annoying git

**Comment by larsbergstrom** [comment]
@bors-servo retry
- servo-mac1 was still being provisioned

**Comment by KiChjang** [comment]
@bors-servo retry
* The annoying git

**Pull request merged**